tests/pipeline/test_pipeline.py

```diff
@@ -238,6 +238,91 @@ def test_data_flow(decam_exposure, decam_reference, decam_default_calibrators, a
+    Does not check measurements, as they do not have the HasBitflagBadness Mixin.
+    """
+    exposure = decam_exposure
+    ref = decam_reference
+    sec_id = ref.section_id
+
+    try:  # cleanup the file at the end
+        p = Pipeline()
+        exposure.badness = 'banding'  # add a bitflag to check for propagation
+
+        # first run the pipeline and check for basic propagation of the single bitflag
+        ds = p.run(exposure, sec_id)
+
+        assert ds.exposure._bitflag == 2     # 2**1 is the bitflag for 'banding'
+        assert ds.image._upstream_bitflag == 2
+        assert ds.sources._upstream_bitflag == 2
+        assert ds.psf._upstream_bitflag == 2
+        assert ds.wcs._upstream_bitflag == 2
+        assert ds.zp._upstream_bitflag == 2
+        assert ds.sub_image._upstream_bitflag == 2
+        assert ds.detections._upstream_bitflag == 2
+        for cutout in ds.cutouts:   # cutouts is a list of cutout objects
+            assert cutout._upstream_bitflag == 2
+
+
+        # test part 2: Add a second bitflag partway through and check it propagates to downstreams
+        # This test 2 is working as implemented here: see my apology in the PR for the mess
+
+        # delete downstreams of ds.sources
+        # i think this session is unnecessary, however I am worried about potentially not cleaning
+        # up properly after the following assignments to None (losing track of the old objects).
+        # not sure if garbage cleaner will get em
+        with SmartSession() as session:
+            # ds.wcs.delete_from_database(session=session)
+            ds.wcs = None
+            # ds.zp.delete_from_database(session=session)
+            ds.zp = None
+            # ds.sub_image.delete_from_disk_and_database(session=session)
+            ds.sub_image = None
+            # probably need a session.commit() somewhere here if I need the commented lins
+
+        ds.sources._bitflag = 2**17  # bitflag 2**17 is 'many sources'
+        desired_bitflag = 2**1 + 2**17 # bitflag for 'banding' and 'many sources'
+        ds = p.run(ds) #ugly solution to cleanup could be make this ds2, then ds2.delete_everything()
+
+        assert ds.wcs._upstream_bitflag == desired_bitflag
+        assert ds.zp._upstream_bitflag == desired_bitflag
+        assert ds.sub_image._upstream_bitflag == desired_bitflag # note my comment in subtraction.py
+                                                                 # only passes if gets bitflag from sources
+
+
+        # test part 3: test update_downstream_badness() function by adding and removing flags
+        # and observing propagation
+        # to Guy: this is partially tested already in test_image.py::test_image_badness,
+        #   but perhaps this is still useful due to testing propagation all the way downstream
+        # ALSO this function will not run on a full ds until get_upstreams has been defined in all the
+        #   downstream classes (currently just sources, image, exposure)
```

Comment:
just drawing attention to this - if I want to run this function on a datastore full of data products, each of them will need `get_upstreams` to be implemented

Reply:
I consider that perhaps implementing get_upstreams on the various objects could be another PR, and finishing this test could be another PR which depends on that

Reply:
nevermind, I will implement `get_upstreams` as part of this PR, as per the meeting discussion